services/galley/src/Galley/API/Update.hs

```diff
@@ -714,6 +716,57 @@ postNewOtrMessage protectee con cnv val msg = do
       gone <- External.deliver toBots
       mapM_ (deleteBot cnv . botMemId) gone
 
+-- | Locally post a message originating from a remote conversation
+-- FUTUREWORK: error handling for missing / mismatched clients
+postRemoteToLocal ::
+  UTCTime ->
+  Remote ConvId ->
+  Qualified (UserId, ClientId) ->
+  Maybe Text ->
+  Maybe Priority ->
+  Bool ->
+  UserClientMap Text ->
+  Galley ()
+postRemoteToLocal now conv sender extra priority transient (UserClientMap rcpts) = do
+  members <- Data.filterRemoteConvMembers (Map.keys rcpts) (unTagged conv)
+  let rcpts' = do
+        m <- members
+        (c, t) <- maybe [] Map.assocs (rcpts ^? ix m)
+        pure (m, c, t)
+  pushSome (map mkPush rcpts')
+  where
+    mkPush :: (UserId, ClientId, Text) -> Push
+    mkPush = remoteToLocalPush now conv sender extra priority transient
+
+remoteToLocalPush ::
+  UTCTime ->
+  Remote ConvId ->
+  Qualified (UserId, ClientId) ->
+  Maybe Text ->
+  Maybe Priority ->
+  Bool ->
+  (UserId, ClientId, Text) ->
+  Push
+remoteToLocalPush now (Tagged conv) sender extra priority transient (rcpt, rcptc, ciphertext) =
+  -- FUTUREWORK: use qualified sender id in message notification
```

Comment:
Can we just do this in this PR? It should be trivial right?

Reply:
Changing the `Event` type is a lot of work, and messes up the diff completely. See for example: https://github.com/wireapp/wire-server/pull/1547